pkg/apis/build/v1alpha1/buildrun_types.go

```diff
@@ -64,8 +64,50 @@ type BuildRunStatus struct {
 	// BuildSpec is the Build Spec of this BuildRun.
 	// +optional
 	BuildSpec *BuildSpec `json:"buildSpec,omitempty"`
+
+	//stages containes the details about the stage for each build with stage start time 
+	// completion and duration
+	// +optional
+	Stages []StageInfo `json:"stages,omitempty"`
+}
+
+// StageInfo contains details about a build stage.
+type StageInfo struct {
```

Comment:
in build v1 / "classic" builds we recently moved away from a custom status type like this toward leveraging the k8s "Condition" pattern .... i.e. something like

To borrow / slightly tweak corev1.PodCondition

```
// various constants of this would be defined
type ConditionType string

type Condition struct {
	// Type is the type of the condition.
	// More info: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#pod-conditions
	Type ConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=PodConditionType"`
	// Status is the status of the condition.
	// Can be True, False, Unknown.

	Status corev1.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=ConditionStatus"`
	// Last time we probed the condition.
	// +optional
	LastProbeTime metav1.Time `json:"lastProbeTime,omitempty" protobuf:"bytes,3,opt,name=lastProbeTime"`
	// Last time the condition transitioned from one status to another.
	// +optional
	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,4,opt,name=lastTransitionTime"`
	// Unique, one-word, CamelCase reason for the condition's last transition.
	// +optional
	Reason string `json:"reason,omitempty" protobuf:"bytes,5,opt,name=reason"`
	// Human-readable message indicating details about last transition.
	// +optional
	Message string `json:"message,omitempty" protobuf:"bytes,6,opt,name=message"`
}
```

I would also contend adding a creation time to that condition type could be useful for our purposes.

In the "learn from recent experiences" category, I think buildv2/k8s builds should start off this way

thoughts @sbose78 @adambkaplan @otaviof ?

Reply:
We should definitely move towards Conditions in our status reporting. The ecosystem has adopted duck-typed Conditions, so using this pattern would unlock a lot of tooling like `kubectl wait` [1].

Upstream has refined their guidance on what conditions types should exist. The more interesting thing to note is the preference towards "abnormal-true" polarity [2]. My interpretation is that "best practice" is to use conditions to surface why something is not working properly. There are, obviously, exceptions to the rule, like what OpenShift does with `ClusterOperator` [3].

[1] https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#wait
[2] https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
[3] https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#what-should-an-operator-report-with-clusteroperator-custom-resource

Reply:
ah ha good references @adambkaplan thanks

Reply:
@adambkaplan @gabemontero thanks for the links. Very useful.

What we do at the moment is actually take the `Tekton` conditions for a particular TaskRun and put some of those conditions into a `BuildRun` Status.

If we do `Conditions` instead of `Status` for the `BuildRun` we will eventually have a `TaskRun` replication into a `BuildRun`, so one of the questions I have is, where do we put the limits in terms of what we copy from a TaskRun? Keep in mind that a `Build` user can get access to the `TaskRun` resources all the time. 

I think my above question also applies to this PR. Opinions?

Reply:
@qu1queee I think replicating Tekton's status is a good start for this PR. And incrementally, we can evolve the `Conditions` reported by build-v2 operator to more meaningful messages.

Using `Conditions` is the "Kubernetes way" of handling change in status reported by an operator, with them is possible to use [`kubectl wait` command](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#wait) to wait for a certain condition to happen, like in our case, a successful image build (`BuildRun`).

So for me is +1 on pursuing the `Conditions`, in short.

Reply:
I will open an issue for this, you guys are so right :)